feature: LVGL display flush offload

danieldegrasse opened the pull request:
Enable offloading of display_write call to background thread for color displays. This feature is opt-in, as it may offer significant performance gains for every display pipeline.

When enabled display_write and lv_disp_flush_ready will be called from a background thread. This means that while the display driver waits on the hardware to render the framebuffer, the LVGL rendering thread will not be blocked.

`CONFIG_LV_Z_FLUSH_THREAD=n`: 72 FPS
`CONFIG_LV_Z_FLUSH_THREAD=y`: 154 FPS


Note: this PR is an implementation of https://github.com/zephyrproject-rtos/lvgl/pull/40 that has been updated now that LVGL gluecode is within Zephyr proper

pdgendt reviewed (dismissed):
I would argue to enable it by default, but it's fine either way.
Thanks!

danieldegrasse commented:
> I would argue to enable it by default, but it's fine either way.
Thanks!

I want to keep this feature as opt in primarily because it won't work with the monochrome display port, where `display_write` can be called twice. For color displays, especially those using LCD controllers, this feature makes sense in most cases.

danieldegrasse committed:
modules: lvgl: allow offloading rendering process to background thread

Enable offloading of display_write call to background thread for color
displays. This feature is opt-in, as it may offer significant
performance gains for every display pipeline.

When enabled display_write and lv_disp_flush_ready will be called from a
background thread. This means that while the display driver waits on the
hardware to render the framebuffer, the LVGL rendering thread will not
be blocked.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: enable LV_Z_FLUSH_THREAD for displays that will benefit from it

Enable LZ_Z_FLUSH_THREAD for display shields and boards that are known
to benefit from this optimization.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented:
@jakkra FYI

jakkra commented:
> @jakkra FYI- this PR has been updated to apply the changes to the LVGL gluecode that is now within the Zephyr tree, and no longer updates the west manifest

Great! One question, where did this go https://github.com/zephyrproject-rtos/lvgl/blob/7b6acfdb03019154500665ce8c919f4d4f88f8e4/zephyr/lvgl_display.c#L110?

danieldegrasse commented:
> Great! One question, where did this go [zephyrproject-rtos/lvgl@7b6acfd/zephyr/lvgl_display.c#L110](https://github.com/zephyrproject-rtos/lvgl/blob/7b6acfdb03019154500665ce8c919f4d4f88f8e4/zephyr/lvgl_display.c#L110)?

Evidently got lost while moving the changes over- glad I pinged you! Should be there now

dleach02 commented on modules/lvgl/Kconfig:
```diff
@@ -53,6 +54,29 @@ endchoice
 
+config LV_Z_FLUSH_THREAD
+	bool "Flush LVGL frames in a separate thread"
+	help
```
Why wouldn't we want to set a default true here for this option?

jakkra replied:
This was asked in the original PR here https://github.com/zephyrproject-rtos/zephyr/pull/59166#issuecomment-1590193173

faxe1008 reviewed (approved):
Thanks for resubmitting against the main repo. Tested this on a mimxrt1060, no issues found.
